frame/vesting/src/lib.rs

```diff
@@ -0,0 +1,610 @@
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Unlock any vested funds of the sender account.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have funds still
+		/// locked under this module.
+		///
+		/// Emits either `VestingCompleted` or `VestingUpdated`.
+		///
+		/// # <weight>
+		/// - `O(1)`.
+		/// - One balance-lock operation.
+		/// - One storage read (codec `O(1)`) and up to one removal.
+		/// - One event.
+		/// # </weight>
+		fn vest(origin) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::update_lock(who)
+		}
+
+		/// Unlock any vested funds of a `target` account.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `target`: The account whose vested funds should be unlocked. Must have funds still
+		/// locked under this module.
+		///
+		/// Emits either `VestingCompleted` or `VestingUpdated`.
+		///
+		/// # <weight>
+		/// - `O(1)`.
+		/// - Up to one account lookup.
+		/// - One balance-lock operation.
+		/// - One storage read (codec `O(1)`) and up to one removal.
+		/// - One event.
+		/// # </weight>
+		fn vest_other(origin, target: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			ensure_signed(origin)?;
+			Self::update_lock(T::Lookup::lookup(target)?)
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// (Re)set or remove the module's currency lock on `who`'s account in accordance with their
+	/// current unvested amount.
+	fn update_lock(who: T::AccountId) -> DispatchResult {
+		ensure!(Vesting::<T>::exists(&who), Error::<T>::NotVesting);
+		let unvested = Self::vesting_balance(&who);
+		if unvested.is_zero() {
+			T::Currency::remove_lock(VESTING_ID, &who);
+			Vesting::<T>::remove(&who);
+			Self::deposit_event(RawEvent::VestingCompleted(who));
+		} else {
+			let reasons = WithdrawReason::Transfer | WithdrawReason::Reserve;
+			T::Currency::set_lock(VESTING_ID, &who, unvested, reasons);
+			Self::deposit_event(RawEvent::VestingUpdated(who, unvested));
+		}
+		Ok(())
+	}
+}
+
+impl<T: Trait> VestingSchedule<T::AccountId> for Module<T> where
+	BalanceOf<T>: MaybeSerializeDeserialize + Debug
+{
+	type Moment = T::BlockNumber;
+	type Currency = T::Currency;
+
+	/// Get the amount that is currently being vested and cannot be transferred out of this account.
+	fn vesting_balance(who: &T::AccountId) -> BalanceOf<T> {
+		if let Some(v) = Self::vesting(who) {
+			let now = <frame_system::Module<T>>::block_number();
+			let locked_now = v.locked_at::<T::BlockNumberToBalance>(now);
+			T::Currency::free_balance(who).min(locked_now)
+		} else {
+			Zero::zero()
+		}
+	}
+
+	/// Adds a vesting schedule to a given account.
+	///
+	/// If there already exists a vesting schedule for the given account, an `Err` is returned
+	/// and nothing is updated.
+	///
+	/// Is a no-op if the amount to be vested is zero.
+	fn add_vesting_schedule(
```

Comment:
what happen if I want to make a sale with vesting schedule to an account with existing lock?
i.e.
Alice balance: $100, all locked by staking module
Alice participant a sale / IPO and have another $100 with a vesting schedule attached.
Alice now have $200 on balance, should have transferable amount of $0
But `T::Currency::set_lock` just add another lock with $100 amount.
Now Alice have $200 on balance, $100 locked with two locks, and $100 transferable.

This is only safe if there was no previous locks. I don't think it is safe to expose this to be used by other modules. More docs helps, but the use case of this method is still going to be very limited.

At least with #4655, it is possible to detect it is unsafe to add vesting schedule and return an error.

Reply:
This is desired behaviour. The idea is that vesting only applies to transferability and that the funds are otherwise usable. If you already had 100 locked for staking or whatever and you bought a pre-locked second 100, then you can think of is at the lock automatically being transferred to the new vesting-locked 100 and freeing up the old pre-vested 100.